toolchain/check/eval.cpp

```diff
@@ -286,37 +286,174 @@ static auto PerformAggregateIndex(Context& context, SemIR::Inst inst)
+
+  auto op = context.insts().GetAs<SemIR::IntLiteral>(arg_id);
+  auto op_val = context.ints().Get(op.int_id);
+
+  if (op_val.isMinSignedValue()) {
+    CARBON_DIAGNOSTIC(CompileTimeIntegerNegateOverflow, Error,
+                      "Integer overflow in negation of {0}.", llvm::APSInt);
+    context.emitter().Emit(loc, CompileTimeIntegerNegateOverflow,
+                           llvm::APSInt(op_val, false));
+  }
+  op_val.negate();
+
+  auto result = context.ints().Add(op_val);
+  return MakeConstantResult(context, SemIR::IntLiteral{op.type_id, result},
+                            Phase::Template);
+}
+
+// Performs a builtin binary integer -> integer operation.
+static auto PerformBuiltinBinaryIntOp(Context& context, SemIRLocation loc,
+                                      SemIR::BuiltinFunctionKind builtin_kind,
+                                      SemIR::InstId lhs_id,
+                                      SemIR::InstId rhs_id)
+    -> SemIR::ConstantId {
+  auto lhs = context.insts().GetAs<SemIR::IntLiteral>(lhs_id);
+  auto rhs = context.insts().GetAs<SemIR::IntLiteral>(rhs_id);
+  auto lhs_val = context.ints().Get(lhs.int_id);
+  auto rhs_val = context.ints().Get(rhs.int_id);
+
+  bool overflow = false;
+  llvm::APInt result_val;
+  llvm::StringLiteral op_str = "<error>";
+  switch (builtin_kind) {
+    case SemIR::BuiltinFunctionKind::IntAdd:
+      result_val = lhs_val.sadd_ov(rhs_val, overflow);
+      op_str = "+";
+      break;
+    case SemIR::BuiltinFunctionKind::IntSub:
+      result_val = lhs_val.ssub_ov(rhs_val, overflow);
+      op_str = "-";
+      break;
+    case SemIR::BuiltinFunctionKind::IntMul:
+      result_val = lhs_val.smul_ov(rhs_val, overflow);
+      op_str = "*";
+      break;
+    case SemIR::BuiltinFunctionKind::IntDiv:
+      if (rhs_val.isZero()) {
+        DiagnoseDivisionByZero(context, loc);
+        return SemIR::ConstantId::Error;
+      }
+      result_val = lhs_val.sdiv_ov(rhs_val, overflow);
+      op_str = "/";
+      break;
+    case SemIR::BuiltinFunctionKind::IntMod:
+      if (rhs_val.isZero()) {
+        DiagnoseDivisionByZero(context, loc);
+        return SemIR::ConstantId::Error;
+      }
+      result_val = lhs_val.srem(rhs_val);
+      // LLVM weirdly lacks `srem_ov`, so we work it out for ourselves:
+      // <signed min> % -1 overflows because <signed min> / -1 overflows.
+      overflow = (lhs_val.isMinSignedValue() && rhs_val.isAllOnes());
+      op_str = "%";
+      break;
+
+    default:
+      CARBON_FATAL() << "Unexpected operation kind.";
+  }
+
+  if (overflow) {
+    CARBON_DIAGNOSTIC(CompileTimeIntegerOverflow, Error,
+                      "Integer overflow in calculation {0} {1} {2}.",
+                      llvm::APSInt, llvm::StringLiteral, llvm::APSInt);
+    context.emitter().Emit(loc, CompileTimeIntegerOverflow,
+                           llvm::APSInt(lhs_val, false), op_str,
+                           llvm::APSInt(rhs_val, false));
```

Comment:
You might consider adding a type conversion for the literal value to APSInt to the diagnostics, rather than writing this in each spot (similar to TypeId -> str). Partly I don't know why "false" (maybe comment with parameter name?)

Reply:
I have an idea here, will send a follow-up PR.